middleware/queue_test.go

```diff
@@ -0,0 +1,235 @@
+package middleware
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestQueueWithConfig(t *testing.T) {
+	e := echo.New()
+
+	handler := func(c echo.Context) error {
+		pts := c.Get("procTime")
+		procTime, ok := pts.(int)
+		if !ok {
+			return c.NoContent(http.StatusInternalServerError)
+		}
+
+		time.Sleep(time.Duration(procTime) * time.Millisecond)
+
+		return c.NoContent(http.StatusOK)
+	}
+
+	mw := QueueWithConfig(QueueConfig{
+		QueueSize:     2,
+		Workers:       1,
+		QueueTimeout:  200 * time.Millisecond,
+		WorkerTimeout: 100 * time.Millisecond,
+	})
+
+	testCases := []struct {
+		procTime int // in Milliseconds
+	}{
+		{50},
+		{95},
+		{95},
+		{95},
+		{95},
+		{120},
+		{250},
+	}
+
+	ch := make(chan int, len(testCases))
+	var wg sync.WaitGroup
+
+	for _, tc := range testCases {
+
+		wg.Add(1)
+
+		go func(pt int) {
+
+			defer wg.Done()
+
+			req := httptest.NewRequest(http.MethodGet, "/", nil)
+			rec := httptest.NewRecorder()
+
+			c := e.NewContext(req, rec)
+			c.Set("procTime", pt)
+
+			err := mw(handler)(c)
+			if err != nil {
+				ch <- err.(*echo.HTTPError).Code
+				return
+			}
+
+			ch <- rec.Code
+
+		}(tc.procTime)
+
+	}
+
+	wg.Wait()
+
+	var errQueueFull, errQueueTimeout, errInternalServerError bool
+
+	for i := 0; i < len(testCases); i++ {
+		c := <-ch
+
+		if c == http.StatusTooManyRequests {
+			errQueueFull = true
+		}
+
+		if c == http.StatusRequestTimeout {
+			errQueueTimeout = true
+		}
+
+		if c == http.StatusInternalServerError {
```

Comment:
This doesn't seems to be useful for testing the middleware code, this error is only returned by the test code (when the procTime is not an int)